# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):
    return x + 2
print (addTwo(4))

#lambda version, without using a variable to pass in data 
print ((lambda x: x + 2)(4))

#using Variables
a_num = 4
print((lambda x: x + 2)(a_num))


6
6
6


#### Saving to a Variable

In [2]:
f_test =lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [3]:
print((lambda x, y, z: x * y * z)(3, 5, 8))
x_test = lambda x, y, z: x * y * z
print(x_test(3, 5, 8))

120
120


#### Passing a Lambda into a Function

In [4]:
def multiply(f, num):
    return f(num)

multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [2]:
#Regular defined function
def Multiiply_tst(num):
    return num * 4

    #function within a function

def returnFunc():
    test = 4 
    def multiply(num):
        return num * 2
    return multiply
f_return = returnFunc()
print(returnFunc)
print(f_return(4))
#Lambda functions returned from a regular function

def returnLam(b,c):
    return lambda x, a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))

<function returnFunc at 0x10cced760>
8
19


#### If Statements within Lambdas

In [3]:
#lambda x: true if condtions else is fals

f_condition = lambda num:num * 2 if num >10 else num + 2

print (f_condition (8))
print (f_condition(12))


10
24


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [4]:
cub_num = 4
print((lambda x: x*x*x)(cub_num))


# def multiply(f, num):
#     return f(num)

# multiply(lambda x: x * x, 4)


# def addTwo(x):
#     return x + 2
# print (addTwo(4))

# #lambda version, without using a variable to pass in data 
# print ((lambda x: x + 2)(4))

# #using Variables
# a_num = 4
# print((lambda x: x + 2)(a_num))


64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [16]:
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2 and num2 ** 2
    else:
        return num, num2
number = [3, 15, 20, 7, 9, 1]
more_nums = [4, 17, 21, 10, 4]

squared_nums_map = list(map(squared, number, more_nums))
print (squared_nums_map)

[16, (15, 17), (20, 21), (7, 10), 16]


#### Using Lambda's with Map

In [17]:
print(list(map(lambda x, y: (x**2, y**2)if x < 10 and y < 10 else (x, y), number, more_nums)))

[(9, 16), (15, 17), (20, 21), (7, 10), (81, 16)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [18]:
print(list(map(lambda x,y: (x*2, y*2)if x <10 and y <10 else(x-1, y-1),number,more_nums)))

[(6, 8), (14, 16), (19, 20), (6, 9), (18, 8)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [ ]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_name(name):
    if name[0].lower() == 'a':
        return True
    elif name[0].lower() =="b":
        return True
    else:
        return False
new_names_list = list(filter(a_name, names))
print(new_names_list)

#### Using Lambda's with Filter()

In [ ]:
new_names_lambda = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lambda)

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [15]:
import statistics
numbers = [4, 11, 20, 3, 15, 20]
def ex_3(number):
    mean = statistics.mean(numbers)
    return number <= mean
new_nums = list(filter(ex_3, numbers))
print(new_nums)

# names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

# def a_name(name):
#     if name[0].lower() == 'a':
#         return True
#     elif name[0].lower() =="b":
#         return True
#     else:
#         return False
# new_names_list = list(filter(a_name, names))
# print(new_names_list)




[4, 11, 3]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [19]:
from functools import reduce

list_1 = [2,4,6,8,10]
def addnums(num1, num2):
    return num1 + num2
result_add = reduce(addnums, list_1)
print (result_add)

def subtract_nums(num1, num2):
    return num1 - num2
result_sub = reduce(subtract_nums, list_1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [20]:
result_lam = reduce(lambda x, y: x+y, list_1)
print(result_lam)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [23]:
list_2 = [1,3,5,7,9]
multiply_lam = reduce(lambda x, y: x*y, list_2)
print(multiply_lam)

945


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [25]:
def add_nums(num):
    if num <= 1:
        print("add_nums(1)=1")
        return num
    else: 
        print(f"add_nums({num}) = {num} + add_nums({num-1})")
        return num + add_nums(num-1)
add_nums(5)

#subtracting one from the variable until the final number is one

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1)=1


15

#### Writing a Factorial Function

In [26]:
def factor_num(num):
    if num <= 1:
        return 1
    else:
        return num + factor_num(num-1)

factor_num(5)

15

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [32]:
# retrun nums - subNums(num - 1)
#subNum(10)

def subtractNum(nums):
    if nums <= 1:
        return nums
    else:
        print(f"Subtracting {nums} by 1 = {nums - 1}")
        return nums - subtractNum(nums -1)
subtractNum(10)

# def add_nums(num):
#     if num <= 1:
#         print("add_nums(1)=1")
#         return num
#     else: 
#         print(f"add_nums({num}) = {num} + add_nums({num-1})")
#         return num + add_nums(num-1)
# add_nums(5)


Subtracting 10 by 1 = 9
Subtracting 9 by 1 = 8
Subtracting 8 by 1 = 7
Subtracting 7 by 1 = 6
Subtracting 6 by 1 = 5
Subtracting 5 by 1 = 4
Subtracting 4 by 1 = 3
Subtracting 3 by 1 = 2
Subtracting 2 by 1 = 1


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [33]:
def my_range(stop,start, step = 2):
    while start < stop:
        yield start
        start += step
for i in my_range(20, start = 2):
    my_generator_value = i
    print (my_generator_value)

my_range(20, start = 2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x10f3fe740>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [35]:
def range_6(stop, start, step = 2):
    while start < stop:
        yield start
        start += step
        for i in range_6(20, start = 2):
            my_gen_value = i
            print(my_gen_value)
range_6 (20, start = 2)

<generator object range_6 at 0x10f4154e0>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [40]:
def unwanted(places,remove):
    return list(filter(lambda x: x not in remove,places))

places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
remove = ["",",","  "," "]

filtered_list = unwanted(places, remove)
print (filtered_list)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [50]:
def sorted_names(author):
    sort_name = []

    for ele in author:
        sort_name.append(ele.split())
    author = []
    for ele in sorted(sort_name, key =lambda x: x[-1].lower()):
        author.append(' '.join(ele))
    return author

author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

print(sorted_names(author))





['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [51]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

farhenheit = list(map(lambda c: (c[0], (9/5) * c[1] + 32),places))

print (farhenheit)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [52]:
def fibonacci(n, current = 0, next_num=1, iteration=0):
    if iteration <= n:
        print(f"Iteration {iteration}: {current}")
        return fibonacci(n, next_num, current + next_num, iteration + 1)
fibonacci(5)

Iteration 0: 0
Iteration 1: 1
Iteration 2: 1
Iteration 3: 2
Iteration 4: 3
Iteration 5: 5
